File: c/resistor-color-trio/src/resistor_color_trio.c
#include "resistor_color_trio.h"

resistor_value_t color_code(resistor_band_t colors[]) {    
    double multiplier = 1;
    double value;
    resistor_unit_t unit;

    for (int i=0; i < (int)colors[2]; ++i) multiplier *= 10;
    value = (10 * (int)colors[0] + (int)colors[1]) * multiplier;

    if (value < 1000) {
        unit = OHMS;
    } else {
        unit = KILOOHMS;
        value /= 1000;
    }
    return (resistor_value_t){value, unit};
}

File: python/resistor-color-trio/resistor_color_trio.py
COLORS = [
        "black",
        "brown",
        "red",
        "orange",
        "yellow",
        "green",
        "blue",
        "violet",
        "grey",
        "white",
        ]

def label(colors):
    codes = [color_code(c) for c in colors]
    value = (codes[0] * 10 + codes[1]) * 10**codes[2]
    if value >= 1_000_000_000:
        return f"{value // 1_000_000_000} gigaohms"
    if value >= 1_000_000:
        return f"{value // 1_000_000} megaohms"
    if value >= 1_000:
        return f"{value // 1_000} kiloohms"
    return f"{value} ohms"

def color_code(color):
    return COLORS.index(color)
